architecture/2024-11-authz-feature-spec.md

```diff
@@ -0,0 +1,390 @@
+
+## User profiles and challenges
+
+**Platform engineers** – Platform engineers configure and maintain Radius. These users are primarily concerned with configuring Radius to support the various development and infrastructure teams within their organization. They need a flexible authorization scheme built into Radius which is easy to setup in the beginning, but flexible enough to grow as Radius is used by more teams in more ways. This flexibility will enable platform engineers to grant and restrict actions at a high level on a wide range of resources, or on a fine-grained level.
+
+**Developers** – Developers use the resource types available to them to model their applications and deploy to environments which they have access to. Some developers will use Radius locally to run their applications, but many will not use Radius directly; instead relying on their organization's CI/CD system to run their application in cloud environments. These developers will not have deep knowledge of Radius beyond using the resource types made available to them. Developers who are using Radius locally should not have to concern themselves with Radius authorization features.
+
+## Scenarios
+
+### Scenario 1 – Getting started and local development
+
+When a first-time user or a developer installs Radius, the authorization configuration should be wide open to the user. The user should never have to configure anything or be presented with any options. This is the behavior of Radius today. There are no enhancements needed to support this scenario.
+
+### Scenario 2 – Simple internal developer platform
+
+A simple internal developer platform is a configuration similar to a first-time setup by a platform engineer for their organization's developers. In this scenario, Radius will be used to manage a handful of applications each with their own development team. A single platform engineering team manages all environments, resource types, and recipes. 
+
+For this scenario, the platform engineer will use the authorization features of Radius to control permissions. However, since Radius is new to the organization, the platform engineer expects to use prescriptive guidance from Radius using pre-defined roles and best practices.
+
+Radius will ship with several roles pre-defined. These roles grant permission to perform actions on certain resource types. These roles will not be assigned to any user out of the box except the Radius administrator role. The documentation will describe how to assign a role to a user or group of users and describe how to define the scope of those permissions.
+
+#### Out-of-the-box roles
+
+**Radius Administrator** – The Radius Administrator has superuser access to the Radius system. 
+
+**Resource Group Administrator** – The Resource Group Administrator role grants permission to manage resource groups, role definitions, and role assignments.
```

Comment:
I agree. I debated this in my mind as well. It could legitimately be separated. Indeed, a Radius admin could modify these groups to be that way. I ended up defaulting to the simpler option which is to group resource groups and roles together.

Reply:
I think this is worth more consideration since you added automatic resource group creation to `rad run` and `rad deploy`.

Reply:
@rynowak Ideally, for scenario 2 (simple enterprise use case), the admin could assign a developer role to a user and that developer would be granted the ability to create resource groups. The developer would have access to that new resource group but not other non-admin users. That does not fit well within our existing model.

I'm assuming this could be implemented with each resource would have an owner property (e.g., developer deploys an app, resource group gets created and the developer assigned as owner), or some other fancy ACL on the resource. That seems complex and not the direction we want to go. 

Thoughts?